diff --git a/gtkhtml/gtkhtml-stream.c b/gtkhtml/gtkhtml-stream.c
index 5cb8cfd..97cb29d 100644
--- a/gtkhtml/gtkhtml-stream.c
+++ b/gtkhtml/gtkhtml-stream.c
@@ -35,6 +35,7 @@ gtk_html_stream_new (GtkHTML *html,
 		     GtkHTMLStreamTypesFunc types_func,
 		     GtkHTMLStreamWriteFunc write_func,
 		     GtkHTMLStreamCloseFunc close_func,
+		     GtkHTMLStreamMIMEFunc mime_func,
 		     gpointer user_data)
 {
 	GtkHTMLStream *new_stream;
@@ -44,6 +45,7 @@ gtk_html_stream_new (GtkHTML *html,
 	new_stream->types_func = types_func;
 	new_stream->write_func = write_func;
 	new_stream->close_func = close_func;
+	new_stream->mime_func = mime_func;
 	new_stream->user_data = user_data;
 
 	return new_stream;
@@ -58,10 +60,29 @@ gtk_html_stream_destroy (GtkHTMLStream *stream)
 }
 
 /**
+ * gtk_html_stream_mime:
+ * @stream: 
+ * @mime_type: mime type or content type
+ *
+ * Set mime type for data in stream.
+ *
+ */
+void
+gtk_html_stream_mime (GtkHTMLStream *stream,
+		       const gchar *mime_type)
+{
+	g_return_if_fail (stream != NULL);
+	g_return_if_fail (mime_type != NULL);
+
+	if (stream->mime_func != NULL)
+		stream->mime_func (stream, mime_type, stream->user_data);
+}
+
+/**
  * gtk_html_stream_write:
- * @stream:
- * @buffer:
- * @size:
+ * @stream: 
+ * @buffer: data
+ * @size: size data to write
  *
  * Write data to a GtkHTMLStream.
  *
@@ -164,6 +185,20 @@ stream_log_types (GtkHTMLStream *stream,
 }
 
 static void
+stream_log_mime (GtkHTMLStream *stream,
+	       const gchar *mime_type,
+	       gpointer user_data)
+{
+	GtkHTMLLog *log = user_data;
+	
+	if (mime_type)
+		fprintf (log->file, "\nSet mime type to %s\n", mime_type);
+
+        gtk_html_stream_mime (log->stream, mime_type);
+
+}
+
+static void
 stream_log_write (GtkHTMLStream *stream,
 	       const gchar *buffer,
 	       size_t size,
@@ -214,6 +249,7 @@ gtk_html_stream_log_new (GtkHTML *html, GtkHTMLStream *stream)
 					  stream_log_types,
 					  stream_log_write,
 					  stream_log_close,
+					  stream_log_mime,
 					  log);
 
 	return new_stream;
diff --git a/gtkhtml/gtkhtml-stream.h b/gtkhtml/gtkhtml-stream.h
index dc7fb53..542bae7 100644
--- a/gtkhtml/gtkhtml-stream.h
+++ b/gtkhtml/gtkhtml-stream.h
@@ -29,6 +29,7 @@ struct _GtkHTMLStream {
 	GtkHTMLStreamWriteFunc write_func;
 	GtkHTMLStreamCloseFunc close_func;
 	GtkHTMLStreamTypesFunc types_func;
+	GtkHTMLStreamMIMEFunc  mime_func;
 	gpointer user_data;
 };
 
@@ -37,10 +38,13 @@ GtkHTMLStream *gtk_html_stream_new       (GtkHTML                *html,
 					  GtkHTMLStreamTypesFunc  type_func,
 					  GtkHTMLStreamWriteFunc  write_func,
 					  GtkHTMLStreamCloseFunc  close_func,
+					  GtkHTMLStreamMIMEFunc   mime_func,
 					  gpointer                user_data);
 void           gtk_html_stream_write     (GtkHTMLStream          *stream,
 					  const gchar            *buffer,
 					  size_t                  size);
+void           gtk_html_stream_mime      (GtkHTMLStream *stream,
+					  const gchar *mime_type);
 void           gtk_html_stream_destroy   (GtkHTMLStream          *stream);
 void           gtk_html_stream_close     (GtkHTMLStream          *stream,
 					  GtkHTMLStreamStatus     status);
diff --git a/gtkhtml/gtkhtml-types.h b/gtkhtml/gtkhtml-types.h
index 400c60c..4eb59ae 100644
--- a/gtkhtml/gtkhtml-types.h
+++ b/gtkhtml/gtkhtml-types.h
@@ -41,6 +41,9 @@ typedef char **(* GtkHTMLStreamTypesFunc) (GtkHTMLStream *stream,
 typedef void   (* GtkHTMLStreamCloseFunc) (GtkHTMLStream *stream,
 					   GtkHTMLStreamStatus status,
 					   gpointer user_data);
+typedef void   (* GtkHTMLStreamMIMEFunc) (GtkHTMLStream *stream,
+					   const gchar *mime_type,
+					   gpointer user_data);
 typedef void   (* GtkHTMLStreamWriteFunc) (GtkHTMLStream *stream,
 					   const gchar *buffer,
 					   size_t size,
diff --git a/gtkhtml/gtkhtml.c b/gtkhtml/gtkhtml.c
index cc168b2..3eae076 100644
--- a/gtkhtml/gtkhtml.c
+++ b/gtkhtml/gtkhtml.c
@@ -3708,6 +3708,22 @@ gtk_html_begin_content (GtkHTML *html, gchar *content_type)
 }
 
 /**
+ * gtk_html_set_mime_type:
+ * @html: the GtkHTML widget the stream belongs to (unused)
+ * @handle: the GkHTMLStream to write to.
+ * @mime_type: stream mime type.
+ *
+ * Set @mime_type for content in @stream.
+ **/
+void
+gtk_html_set_mime_type (GtkHTML *html,
+		GtkHTMLStream *handle,
+		const gchar *mime_type)
+{
+	gtk_html_stream_mime (handle, mime_type);
+}
+
+/**
  * gtk_html_write:
  * @html: the GtkHTML widget the stream belongs to (unused)
  * @handle: the GkHTMLStream to write to.
@@ -4663,13 +4679,25 @@ gtk_html_get_default_engine(GtkHTML *html)
 	return html_engine_get_engine_type( html->engine);
 }
 
-
+/**
+ * gtk_html_set_default_content_type:
+ * @html: the GtkHTML widget
+ * @content_type: default content for main page.
+ *
+ * Set default @content_type for main page in @html widget.
+ **/
 void
 gtk_html_set_default_content_type (GtkHTML *html, const gchar *content_type)
 {   
     html_engine_set_content_type( html->engine, content_type);
 }
 
+/**
+ * gtk_html_get_default_content_type:
+ * @html: the GtkHTML widget
+ *
+ * Returns: default content_type for main page in @html widget.
+ **/
 const gchar*
 gtk_html_get_default_content_type (GtkHTML *html)
 {   
diff --git a/gtkhtml/gtkhtml.h b/gtkhtml/gtkhtml.h
index 0120c76..535e446 100644
--- a/gtkhtml/gtkhtml.h
+++ b/gtkhtml/gtkhtml.h
@@ -165,6 +165,9 @@ void                       gtk_html_write                         (GtkHTML
 								   GtkHTMLStream             *handle,
 								   const gchar               *buffer,
 								   size_t                     size);
+void                       gtk_html_set_mime_type                 (GtkHTML                   *html,
+								   GtkHTMLStream             *handle,
+								   const gchar               *mime_type);
 void                       gtk_html_end                           (GtkHTML                   *html,
 								   GtkHTMLStream             *handle,
 								   GtkHTMLStreamStatus        status);
diff --git a/gtkhtml/htmlengine.c b/gtkhtml/htmlengine.c
index 833f6fe..e35eb75 100644
--- a/gtkhtml/htmlengine.c
+++ b/gtkhtml/htmlengine.c
@@ -116,6 +116,9 @@ static void      html_engine_stream_write     (GtkHTMLStream       *stream,
 					       const gchar         *buffer,
 					       size_t               size,
 					       gpointer             data);
+static void      html_engine_stream_mime      (GtkHTMLStream *handle,
+					       const gchar *mime_type,
+					       gpointer data);				       
 static void      html_engine_stream_end       (GtkHTMLStream       *stream,
 					       GtkHTMLStreamStatus  status,
 					       gpointer             data);
@@ -4663,6 +4666,7 @@ html_engine_begin (HTMLEngine *e, char *content_type)
 					  html_engine_stream_types,
 					  html_engine_stream_write,
 					  html_engine_stream_end,
+					  html_engine_stream_mime,
 					  e);
 #ifdef LOG_INPUT
 	if (getenv("GTK_HTML_LOG_INPUT_STREAM") != NULL)
@@ -4712,6 +4716,22 @@ html_engine_stream_types (GtkHTMLStream *handle,
 }
 
 static void
+html_engine_stream_mime (GtkHTMLStream *handle,
+			  const gchar *mime_type,
+			  gpointer data)
+{
+	HTMLEngine *e;
+
+	e = HTML_ENGINE (data);
+
+	if (mime_type == NULL)
+		return;
+
+	html_engine_set_content_type (e, mime_type);
+
+}
+
+static void
 html_engine_stream_write (GtkHTMLStream *handle,
 			  const gchar *buffer,
 			  size_t size,
diff --git a/gtkhtml/htmlengine.h b/gtkhtml/htmlengine.h
index 9ef27d9..3c3bd9e 100644
--- a/gtkhtml/htmlengine.h
+++ b/gtkhtml/htmlengine.h
@@ -318,7 +318,7 @@ void           html_engine_flush            (HTMLEngine  *e);
 void           html_engine_set_engine_type   (HTMLEngine *e,
 					 gboolean engine_type);
 gboolean       html_engine_get_engine_type   (HTMLEngine *e);
-void 		   html_engine_set_content_type(HTMLEngine *e,
+void           html_engine_set_content_type(HTMLEngine *e,
 					const gchar* content_type);
 const gchar *  html_engine_get_content_type(HTMLEngine *e);
 
diff --git a/gtkhtml/htmlimage.c b/gtkhtml/htmlimage.c
index 2e105c1..fe5ac06 100644
--- a/gtkhtml/htmlimage.c
+++ b/gtkhtml/htmlimage.c
@@ -1418,6 +1418,7 @@ html_image_pointer_load (HTMLImagePointer *ip)
 				    html_image_factory_types,
 				    html_image_factory_write_pixbuf,
 				    html_image_factory_end_pixbuf,
+				    NULL, /* original image type not not needed */
 				    ip);
 }
 
diff --git a/gtkhtml/testgtkhtml.c b/gtkhtml/testgtkhtml.c
index c0de89b..3fa59d1 100644
--- a/gtkhtml/testgtkhtml.c
+++ b/gtkhtml/testgtkhtml.c
@@ -666,7 +666,7 @@ got_data (SoupSession *session, SoupMessage *msg, gpointer user_data)
 	ContentType = soup_message_headers_get (msg->response_headers, "Content-type");
 
 	if (ContentType != NULL)
-		gtk_html_set_default_content_type (html, ContentType);
+		gtk_html_set_mime_type (html, handle, ContentType);
 
 	gtk_html_write (html, handle, msg->response_body->data,
 			msg->response_body->length);
