Index: components/Makefile.am
===================================================================
--- components/Makefile.am	(revision 9161)
+++ components/Makefile.am	(working copy)
@@ -1,5 +1,2 @@
-if BONOBO_EDITOR
-BONOBO_EDITOR_SUBDIR = html-editor
-endif
 
-SUBDIRS = editor $(BONOBO_EDITOR_SUBDIR)
+SUBDIRS = editor
Index: configure.in
===================================================================
--- configure.in	(revision 9161)
+++ configure.in	(working copy)
@@ -229,7 +229,6 @@
 gtkhtml/libgtkhtml.pc
 a11y/Makefile
 components/Makefile
-components/html-editor/Makefile
 components/editor/Makefile
 components/editor/gtkhtml-editor.pc
 art/Makefile
Index: testgtkhtml.c
===================================================================
--- gtkhtml/testgtkhtml.c	(revision 9161)
+++ gtkhtml/testgtkhtml.c	(working copy)
@@ -1053,6 +1053,9 @@
 
 	session = soup_session_async_new ();
 
+	SoupCookieJar * cookie_jar = soup_cookie_jar_text_new("./cookies.txt", FALSE);
+	soup_session_add_feature(session, SOUP_SESSION_FEATURE(cookie_jar));
+
 	html_widget = gtk_html_new ();
 	html = GTK_HTML (html_widget);
 	gtk_html_set_allow_frameset (html, TRUE);
